### PR TITLE
Feature: boundary condition handler MPI communication

### LINKED ISSUE #2405: Activate boundary conditions for broadcasted, derived, etc. cmeshes.
The boundary condition handler should be able to communicated the hash maps across all ranks and also be copied to other new cmeshes.

### PR DESCRIPTION
**Pull request by sandro-elsweijer**
**_Describe your changes here:_**
The boundary condition handler implemented in #2401 only works for new cmeshes which are created equally on all processes.
This PR introduces a bcast and synchronize function, which can be used for t8_cmesh_bcast or for partitioned cmeshes.
Closes #2405 

**_All these boxes must be checked by the AUTHOR before requesting review:_**
- [x] The PR is *small enough* to be reviewed easily. If not, consider splitting up the changes in multiple PRs.
- [x] The title starts with one of the following prefixes: `Documentation:`, `Bugfix:`, `Feature:`, `Improvement:` or `Other:`.
- [x] If the PR is related to an issue, make sure to link it.
- [x] The author made sure that, as a reviewer, he/she would check all boxes below.

**_All these boxes must be checked by the REVIEWERS before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is *fully understood, bug free, well-documented and well-structured*.
#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually.
- [ ] The code follows the [t8code coding guidelines](https://github.com/DLR-AMR/t8code/wiki/Coding-Guideline).
- [ ] New source/header files are properly added to the CMake files.
- [ ] The code is well documented. In particular, all function declarations, structs/classes and their members have a proper doxygen documentation. Make sure to add a file documentation for each file!
- [ ] README.md files are updated if necessary.
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue).
#### Tests
- [ ] The code is covered in an existing or new test case using Google Test.
- [ ] The code coverage of the project (reported in the CI) should not decrease. If coverage is decreased, make sure that this is reasonable and acceptable.
- [ ] Valgrind doesn't find any bugs in the new code. [This script](https://github.com/DLR-AMR/t8code/blob/main/scripts/check_valgrind.sh) can be used to check for errors; see also this [wiki article](https://github.com/DLR-AMR/t8code/wiki/Debugging-with-valgrind).

If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually).
#### Scripts and Wiki
- [ ] If a new directory with source files is added, it must be covered by the `scripts/internal/find_all_source_files.sh` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example or tutorial and a Wiki article.
#### License
- [ ] The author added a BSD statement to `doc/` (or already has one).

### CODE REVIEW

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.hxx [hunk 1]
```diff
@@ -267,6 +371,45 @@ struct t8_cmesh_boundary_condition_handler
...
+  serialize_map () const;
+
+  /**
+   * Unpacks and integrates the \a serial_data into m_boundary_conditions.
+   * It either merges the already existing data or overwrites the complete map if \a overwrite is set to true.
+   * \param [in]  serial_data   The data to unpack and integrate.
+   * \param [in]  overwrite     Overwrites the data in this handler if true. Merges the data with the existing data on false.
+   */
+  void
+  unpack_map (std::vector<char> &serial_data, bool overwrite);
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.hxx]:
```suggestion
  unpack_map (const std::vector<char> &serial_data, const bool overwrite);
```

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx [hunk 2]
```diff
@@ -91,3 +94,103 @@ t8_cmesh_boundary_condition_handler::get_boundary_condition_attribute_key () con
...
+  return serial_data;
+}
+
+void
+t8_cmesh_boundary_condition_handler::unpack_map (std::vector<char> &serial_data, bool overwrite)
+{
+  if (overwrite) {
+    m_boundary_conditions.clear ();
+  }
+
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx]: The loop is neat!
I only have 2 concers, which can be fixed quickly
- If there is no data (which might happen), then serial_data.data () + serial_data.size () does nullpointer arithmetic which is often not allowed.

Solution: Resolve this by checking if serial_data.size () is 0 and skipping the reading.
 
- The construct relies on serial_data being separated by \0 and especially ending in \0. Since it comes from MPI communication there might be communication errors and we might loose a trailing \0 or parts.
If that would be the case, then we are starting to access memory regions that are not ours here.

Solution: Check that the last byte in serial_data is actually \0 (not assert but check even in non-debug mode).

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx [hunk 3]
```diff
@@ -91,3 +94,103 @@ t8_cmesh_boundary_condition_handler::get_boundary_condition_attribute_key () con
...
+    /* Nothing to do. */
+    return;
+  }
+
+  /* Prepare the sendbuffer. */
+  std::vector<char> send_buffer = t8_cmesh_boundary_condition_handler::serialize_map ();
+
+  /* Communicate the local sizes */
+  int local_size = static_cast<int> (send_buffer.size ());
+  std::vector<int> sizes (mpisize);
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx]: This creates a vector of size mpisize with the same entries on each process.
This is a large amount of memory when mpisize is large and it grows quadratically with mpisize

Please use MPI shared memory in this case with a t8_shmem array.
This way, only one array per shared memory region will be allocated.

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx [hunk 4]
```diff
@@ -91,3 +94,103 @@ t8_cmesh_boundary_condition_handler::get_boundary_condition_attribute_key () con
...
+  sc_MPI_Allgather (&local_size, 1, sc_MPI_INT, sizes.data (), 1, sc_MPI_INT, comm);
+
+  /* Compute the offsets for the data and create receive buffer. */
+  std::vector<int> offsets (mpisize);
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx]: See my comment on sizes  above

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx [hunk 5]
```diff
@@ -91,3 +94,103 @@ t8_cmesh_boundary_condition_handler::get_boundary_condition_attribute_key () con
...
+  for (int rank = 0; rank < mpisize; ++rank) {
+    offsets[rank] = total_size;
+    total_size += sizes[rank];
+  }
+  std::vector<char> recv_buffer (total_size);
+
+  /* Communicate. */
+  sc_MPI_Allgatherv (send_buffer.data (), local_size, sc_MPI_BYTE, recv_buffer.data (), sizes.data (), offsets.data (),
+                     sc_MPI_BYTE, comm);
+
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.cxx]: I believe we could get duplicate values in the recv_buffer.

If the cmesh has shared trees, then the boundary condition info of these shared trees will be present on multiple processes and land as duplicates in the recv_buffer.

Might this also happen during setup when users do not care for the process partitions when assigning boundary conditions?

If you agree that duplicates might occur, then the unpack_map function should check for duplicates and remove them.

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.hxx [hunk 6]
```diff
@@ -267,6 +371,45 @@ struct t8_cmesh_boundary_condition_handler
     return m_boundary_conditions.at (hash);
   }
 
+  /**************************************** MPI HELPER FUNCTIONS ****************************************/
+
+ public:
+  /**
+   * Synchronizes the contents of the boundary condition handler across all processes.
+   * \param [in]  comm      The communicator to use.
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.hxx]:
```suggestion
   * \param [in]  comm      The communicator to use.
   * \note This is an MPI collective function and must be called by all processes of \a comm.
```

File: src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.hxx [hunk 7]
```diff
@@ -267,6 +371,45 @@ struct t8_cmesh_boundary_condition_handler
...
+   */
+  void
+  synchronize (sc_MPI_Comm comm);
+
+  /**
+   * Broadcasts the boundary conditions from \a main_rank to all other ranks.
+   * \param [in]  main_rank The main rank from which to broadcast.
+   * \param [in]  comm      The communicator to use.
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_boundary_conditions/internal/t8_cmesh_boundary_condition_handler.hxx]:
```suggestion
   * \param [in]  comm      The communicator to use.
   * \note This is an MPI collective function and must be called by all processes of \a comm.
```

File: src/t8_cmesh/t8_cmesh_internal/t8_cmesh_commit.cxx [hunk 8]
```diff
@@ -503,6 +504,20 @@ t8_cmesh_commit_partitioned_new (t8_cmesh_t cmesh, sc_MPI_Comm comm)
...
 
+  /* Communicate the boundary conditions.
+   * Since some processes can have no trees,
+   * we have to communicate if boundary conditions were applied first. */
+  int boundary_conditions_applied_locally = cmesh->boundary_condition_handler != nullptr;
+  int boundary_conditions_applied_globally = 0;
+  sc_MPI_Allreduce (&boundary_conditions_applied_locally, &boundary_conditions_applied_globally, 1, sc_MPI_INT,
+                    sc_MPI_MAX, comm);
+  if (boundary_conditions_applied_globally && cmesh->boundary_condition_handler == nullptr) {
+    t8_cmesh_add_boundary_condition_handler (cmesh);
```
Comment (holke) [inline src/t8_cmesh/t8_cmesh_internal/t8_cmesh_commit.cxx]: Add a check that this was succesfull (i.e. not nullptr returned).

File: test/t8_cmesh/t8_gtest_cmesh_boundary_conditions.cxx [hunk 9]
```diff
@@ -315,12 +338,226 @@ TEST (t8_gtest_cmesh_boundary_conditions, test_boundary_condition_c_interface)
...
+   * has 2 shared bcs with rank - 1, 2 unique bcs and 2 shared bcs with rank + 1.
+   * Rank 0 and mpisize - 2 have 4 unique bcs and 2 shared ones (if there are more than 2 ranks).
+   * Rank mpisize - 1 has no boundary conditions. */
+  std::vector<std::string> testing_boundary_conditions;
+  const size_t num_boundary_conditions_per_rank = 6;
+  /* Since the last rank gets no bcs we use mpisize - 1 and since every rank gets 2 additional bcs we also add 2. */
+  size_t num_boundary_conditions = (mpisize - 1) * num_boundary_conditions_per_rank + 2;
+  /* If we are serial we just apply 6 boundary conditions. */
+  if (mpisize == 1)
+    num_boundary_conditions = num_boundary_conditions_per_rank;
```
Comment (holke) [inline test/t8_cmesh/t8_gtest_cmesh_boundary_conditions.cxx]:
```suggestion
  if (mpisize == 1)
  {
    num_boundary_conditions = num_boundary_conditions_per_rank;
  }
```
Just coding conventions.

**Review by holke — CHANGES_REQUESTED** [review]
Very nice addition, good implementation and well tested :)
I only have some minor remarks in the code and one question here:
 
- In the t8_cmesh_boundary_condition_handler struct you are mixing private and public around all the time. Instead of defining all public functions first and private functions after as usual. What are your reasons for doing it this way?
